### PR TITLE
Earth - Lina

### PR DESCRIPTION
**Pull request by lina5147**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | A custom method I wrote is the top_10 class method that takes in the media category as the argument. Depending on the category, it collects all of the work with that specific category. It first sorts the work by vote and if there are ties in votes, it orders it alphabetically and then returns the first 10 works.
Describe how you approached testing that model method. What edge cases did you come up with? | I used fixtures to help generate works, users, and votes. I first tested nominal cases such as, extracting 10 works of the category, extract works even if there are less than 10, list top 10 by number of votes. My edge cases include listing works even if there are no votes, list 0 works if the category does not have any works.
What are `session` and `flash`? What is the difference between them? | Session is used to help keep track of data. It is a hash-like-object. The data store will be stored indefinitely until the user closes out of the browser or manually remove the data. On the other hand, flash is used to send one-time messages. It goes away after the next response cycle. It is good for sending success or failure messages to the view.
What was one thing that you gained more clarity on through this assignment? | I gained more clarity on relations, especially one to many relations. Understanding how to access certain data bases on relationships between model.
What is the Heroku URL of your deployed application? |  https://fathomless-depths-59502.herokuapp.com/

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+
+  root to: 'homepages#index'
+
+  resources :works do
+    resources :votes, only: [:create]
+  end
+
+  resources :votes, only: [:create]
```
Comment (CheezItMan) [inline config/routes.rb:10]: You already have the nested route, this isn't needed.

```suggestion
```

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+          <td width="100">
+            <%= work.publication_year%>
+          </td>
+          <td width="100">
+            <%= button_to "Upvote", work_votes_path(work.id) %>
+          </td>
+        </tr>
+      <% end %>
+      </tbody>
+    </table>
```
Comment (CheezItMan) [inline app/views/works/index.html.erb:37]: Notice that you do this kind of table a **LOT**!  Instead of repeating this logic in your views over and over again, you can create a partial and pass to the partial a collection of Work instances.  That way you can dry up your code a bit.

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  has_many :votes
+
+  def self.top_10(category)
+
+    # works by category
+    works_by_category = self.where(category: category)
+    # top 10 order by number of votes
+    # ties? alphabetical order by title
+    return works_by_category.sort_by { |x| [-x.votes.count, x.title]}.first(10)
```
Comment (CheezItMan) [inline app/models/work.rb:13]: Just noting that you can also use postgres and the `.order` method to sort items by the number of votes.  Just FYI, that's best-practice as the DB is more efficient at this than Ruby.

This is because if the DB sorts the items and selects just 10 to send to Rails, that's a LOT less records to transmit to Rails.

File: test/models/work_test.rb [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+      expect(result).must_equal true
+    end
+
+    it 'is invalid without a title' do
+      @work.title = nil
+
+      result = @work.valid?
+
+      expect(result).must_equal false
```
Comment (CheezItMan) [inline test/models/work_test.rb:37]: You should also have a check for the error message.  You know it's not valid, but not necessarily why.

File: test/models/work_test.rb [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+          title: "test work",
+          creator: "lalala",
+          publication_year: 2010,
+          description: "blah blah"
+      )
+
+      result = @work.valid?
+
+      expect(result).must_equal false
+      expect(@work.errors.messages).must_include :title
```
Comment (CheezItMan) [inline test/models/work_test.rb:53]: You should also have a check for the error message.  You know it's not valid, but not necessarily why.

File: test/models/work_test.rb [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    end
+
+    it 'list top 10 by number of votes' do
+
+      category = "book"
+
+      books = Work.top_10(category)
+
+      expect(books.first).must_equal works(:book1)
+      expect(books.first.votes.count).must_equal 5
```
Comment (CheezItMan) [inline test/models/work_test.rb:117]: I also suggest testing that each item has greater than or equal to the number of votes of subsequent works.

File: test/models/work_test.rb [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+    end
+
+    it 'list top 10 in alphabetical order for ties in votes' do
```
Comment (CheezItMan) [inline test/models/work_test.rb:121]: 👍   Good check

File: test/controllers/homepages_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe HomepagesController do
```
Comment (CheezItMan) [inline test/controllers/homepages_controller_test.rb:3]: Just noting no tests

File: test/controllers/users_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+    it "can login an existing user" do
+      user = User.create(username: "Bobbee Fulter")
+
+      # if user exist, should not create a new user
+      expect {
+        login(user.username)
+      }.wont_change "User.count"
+
+      expect(flash[:success]).wont_be_nil
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb:37]: You can also test for the content of the flash, so that you know if it's saying it's a new user or welcoming back an existing user.

File: test/controllers/votes_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe VotesController do
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:3]: Just noting no tests

**Review by CheezItMan — COMMENTED** [review]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ⚠️, I get an error page, broken view
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️, one small comment
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ⚠️ Minor Note you can use a partial instead of repeating your list of works over and over again in many views. 
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ⚠️ No controller filters
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ⚠️ No controller filters
| 12. The `WorksController` uses strong params | ⚠️ No controller filters
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️, very well done

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | 💚





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

## Summary

Outstsanding work Lina.   You hit the learning goals here.  I made some minor notes, but this is outstanding work.  I like the styling and model methods especially.  

The only area I have for improvement is for you to practice using controller filters.  Otherwise great!
